lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -428,93 +433,165 @@ static SILValue emitCodeForSymbolicValue(SymbolicValue symVal,
-    continueLifetimeEndInstructionSearch = false;
-
-    for (Operand *use : currValue->getUses()) {
-      SILInstruction *user = use->getUser();
+/// Collect the end points of the instructions that are data dependent on \c
+/// value. A instruction is data dependent on \c value if its result may
+/// transitively depends on \c value. Note that data dependencies through
+/// addresses are not tracked by this function.
+///
+/// \param value SILValue that is not an address.
+/// \param fun SILFunction that defines \c value.
+/// \param endUsers buffer for storing the found end points of the data
+/// dependence chain.
+static void
+getEndPointsOfDataDependentChain(SILValue value, SILFunction *fun,
+                                 SmallVectorImpl<SILInstruction *> &endUsers) {
+  assert(!value->getType().isAddress());
+
+  // Collect the instructions that are data dependent on the value using a
+  // fix point iteration.
+  SmallPtrSet<SILInstruction *, 16> visitedUsers;
+  SmallVector<SILInstruction *, 16> worklist;
+  // Initialize worklist with the uses of value. Note that uses should not
+  // contain value's defining instruction.
+  for (Operand *use : value->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (visitedUsers.insert(user).second)
+      worklist.push_back(user);
+  }
 
-      if (isa<ReleaseValueInst>(user) || isa<DestroyValueInst>(user)) {
-        lifetimeEndInsts.push_back(user);
-        continue;
+  while (!worklist.empty()) {
+    SILInstruction *currInst = worklist.pop_back_val();
+    for (SILValue result : currInst->getResults()) {
+      for (Operand *use : result->getUses()) {
+        SILInstruction *user = use->getUser();
+        if (visitedUsers.insert(user).second)
+          worklist.push_back(user);
       }
+    }
+  }
 
-      if (isa<CopyValueInst>(user)) {
-        auto *copyValueInst = cast<CopyValueInst>(user);
-        // Continue looking for the end-of-lifetime instruction for the
-        // result of copy_value.
-        currValue = copyValueInst;
-        continueLifetimeEndInstructionSearch = true;
-      }
+  // At this point, visitedUsers have all the transitive, data-dependent uses.
+  // Compute the lifetime frontier of all the uses which are the instructions
+  // following the last uses. Every exit from the last uses will have a
+  // lifetime frontier.
+  SILInstruction *valueDefinition = value->getDefiningInstruction();
+  SILInstruction *def =
+      valueDefinition ? valueDefinition : &(value->getParentBlock()->front());
+  ValueLifetimeAnalysis lifetimeAnalysis =
+      ValueLifetimeAnalysis(def, SmallVector<SILInstruction *, 16>(
+                                     visitedUsers.begin(), visitedUsers.end()));
+  ValueLifetimeAnalysis::Frontier frontier;
+  bool hasCriticlEdges = lifetimeAnalysis.computeFrontier(
+      frontier, ValueLifetimeAnalysis::DontModifyCFG);
+  endUsers.append(frontier.begin(), frontier.end());
+  if (!hasCriticlEdges)
+    return;
```

Comment:
It would have been fine to avoid all the critical edges complexity and just split critical edges in the beginning, as other passes do. Eventually we'll just disallow them everywhere so we'll need to come back and cleanup code like this.